File: old_python_judgement/util/player.py
import random

class Player:
    """
    This is the general Player class. It implements the random AI.
    All classes that extend this can only overwrite the predict and choose
    function.
    """
    def __init__(self, name): 
        self.hand = []
        self.name = name
        self.won = 0
    
    def getLegalMoves(self, cardsPlayed):
        if cardsPlayed == []:
            return self.hand
        firstSuit = cardsPlayed[0].suit
        cardsOfSuit = [card for card in self.hand if card.suit == firstSuit]
        if cardsOfSuit == []:
            return self.hand
        return cardsOfSuit
    
    def win(self):
        self.won += 1


    def predict(self, numCards, numLeft, curPredictions, trump, last = False):
        if not last:
            self.prediction = random.randint(1,numCards)
            return self.prediction
        else:
            self.prediction = random.randint(1,numCards)
            while self.prediction == numLeft:
                self.prediction = random.randint(1,numCards)
            return self.prediction



    def choose(self, cardsPlayed, trump):
        moves = self.getLegalMoves(cardsPlayed)
        card = random.choice(moves)
        self.hand.remove(card)
        return card

    def reset(self):
        self.won = 0


class HumanPlayer(Player):
    def predict(self, numCards, numLeft, curPredictions, trump, last = False):
        print("Current predictions are: " + str(curPredictions))
        print("Your hand is: " + str(self.hand))
        print("What do you predict to make?")
        self.prediction = int(input("Prediction: "))
        return self.prediction

    def choose(self, cardsPlayed, trump):
        legalMoves = self.getLegalMoves(cardsPlayed)
        print("Cards Played So far: " + str(cardsPlayed))
        print("Legal Cards in your hand: " + str(legalMoves))
        card = legalMoves[int(input("Location of card you want to play: \n")) - 1]
        self.hand.remove(card)
        return card

class AggressiveBot(Player):
    def numTrump(self, trump):
        return len([card for card in self.hand if card.suit == trump])

    def predict(self, numCards, numLeft, curPredictions, trump, last = False):
        self.prediction = self.numTrump(trump.suit)
        return self.prediction

    def choose(self, cardsPlayed, trump):
        moves = self.getLegalMoves(cardsPlayed)
        if self.won <= self.prediction:
            for card in moves:
                if self.wouldWinCurHand(cardsPlayed + [card], card,trump.suit):
                    self.hand.remove(card)
                    return card
            card = random.choice(moves)
            self.hand.remove(card)
            return card
        else:
            for card in moves:
                if not self.wouldWinCurHand(cardsPlayed + [card], card,trump.suit):
                    self.hand.remove(card)
                    return card
            card = random.choice(moves)
            self.hand.remove(card)
            return card


    def wouldWinCurHand(self, cardsPlayed, card, trumpSuit):
        card1 = cardsPlayed[0]
        winner = card1
        loc = 0
        trumpSeen = winner.suit == trumpSuit
        for i in range(1,len(cardsPlayed)):
            compCard = cardsPlayed[i]
            if winner.suit == compCard.suit and winner.rank < compCard.rank:
                winner = compCard
            elif compCard.suit == trumpSuit and not trumpSeen:
                winner = compCard
        return winner==card

File: game_server/src/objects/player.ts
import { randomBytes } from "crypto";
import { Card } from "./card";
import _ from "lodash";

// This will be class that defines all player available actions
export class RegulatedPlayer {
  id: number;
  hand: Card[];
  otherPlayerIds: number[];

  constructor(id: number) {
    this.id = id;
    this.hand = [];
    this.otherPlayerIds = [];
  }

  getAllPlayer(ids: number[]){
    this.otherPlayerIds = ids;
  }

  getDealtCards(cards: Card[]) {
    cards.map((card) => this.hand.push(card));
  }

  getLegalMoves(cardsPlayedInOrder: Card[]): Card[] {
    if (cardsPlayedInOrder.length === 0) {
      return this.hand;
    }
    const firstSuit = cardsPlayedInOrder[0].suit;
    let legalCards = this.hand.filter((card) => card.suit === firstSuit);
    if (legalCards.length === 0) {
      return this.hand;
    }
    return legalCards;
  }

  playCard(cardsPlayedInOrder: Card[], trumpCard: Card) {
    const legalCards = this.getLegalMoves(cardsPlayedInOrder);
    const randomChoice = _.random(0, legalCards.length - 1, false);
    const cardToPlay = legalCards[randomChoice];
    const indexToPopOff = this.hand.findIndex((card) => card === cardToPlay);
    return this.hand.splice(indexToPopOff, 1)[0];
  }

  // Logic for this is broken, pass in Predictions So Far and Trump Card only.
  predict(numCards: number, numLeft: number, trumpCard: Card, last: boolean = false): number {
    let prediction = _.random(0, numCards, false);
    if (last && prediction === numLeft) {
      prediction = prediction + [-1, 1][_.random(0, 1, false)];
    }
    return prediction;
  }
}

export class Player extends RegulatedPlayer {
  playCard(cardsPlayedInOrder: Card[], trumpCard: Card) {
    const legalCards = this.getLegalMoves(cardsPlayedInOrder);
    const randomChoice = _.random(0, legalCards.length - 1, false);
    const cardToPlay = legalCards[randomChoice];
    const indexToPopOff = this.hand.findIndex((card) => card === cardToPlay);
    return this.hand.splice(indexToPopOff, 1)[0];
  }

  // Logic for this is broken, pass in Predictions So Far and Trump Card only.
  predict(numCards: number, numLeft: number, trumpCard: Card, last: boolean = false): number {
    let prediction = _.random(0, numCards, false);
    if (last && prediction === numLeft) {
      prediction = prediction + [-1, 1][_.random(0, 1, false)];
    }
    return prediction;
  }
}
